# Name: bgenie_output_filter.sh
# Date: 15.01.2019, edited 06.05.2019
# Author: Amaia Carrion Castillo
# Description: Filters out BGENIE output to include only SNPs QCtool INFO>0.7 (from total UKB sample), and HWE p>1e-7. The goal is to avoid doing this SNP filtering per phenotype, and make the plotting/cleaning part run in R less memory heavy.
# Dependencies:
## association_bgenie_impUKBv3_*.sh # runs the association and creates input files for the current script.
## ?

#----------------------------------------------------------------------
#root=ukb25465_ukb25468
#subset_name=imagingT1_N18057
#type=Volume
#region=Planum_Temporale_totalBV
root=${1}
subset_name=${2}
type=${3}
region=${4}
#----------------------------------------------------------------------

#----------------------------------------------------------------------
resource_dir=/data/workspaces/lag/shared_spaces/Resource_DB/
analysis_dir=/data/workspaces/lag/workspaces/lg-ukbiobank/analysis/amaia/genetic_analysis/release_v3/
working_dir=/data/workspaces/lag/workspaces/lg-ukbiobank/working_data/amaia/genetic_data/release_v3/imp/bgenie/output/PT/${subset_name}/
working_dir2=/data/clusterfs/lag/users/amacar/ukb/output/
qc_dir=/data/workspaces/lag/workspaces/lg-ukbiobank/working_data/amaia/genetic_data/release_v3/imp/subset_${subset_name}/QC/
#----------------------------------------------------------------------
maf_thr=0.001
info_thr=0.7
hwe_thr=1e-07

#----------------------------------------------------------------------
cd ${working_dir}
#----------------------------------------------------------------------
echo "Alternate_id RS_id Position Allele1 Allele2 MAF Minor_Allele Info_score" > ${working_dir}mfi_header.tmp

for i in {1..22} X
 do
  f=${root}_${subset_name}_bgenie_${type}_${region}_chr${i}.out.gz
  #f=${root}_${subset_name}_bgenie_chr${i}.out.gz
  if [ -f ${f} ]
   then
    f2=$(echo ${f} | sed 's/.gz//g')
    # extract chromosome info from the filename
    chr=chr$(echo ${f} | awk -F'chr' '{print $2}' | sed 's/.out.gz//g')
    # list of SNPs to keep from the QC file generated by snp_qc_subset_imaging_imp.R
    f_qc2keep=$(ls ${qc_dir}/*${chr}_*_snp-stats*_mfi_hrc.snps2keep)
    # define output file
    # define output file
    f_out=$(echo ${f} | sed "s/.out.gz/_'${hwe_thr}'HWEp_'${info_thr}'INFO_'${maf_thr}'MAF.out.gz/g")
    f_out=$(echo ${f_out} | sed "s/'//g")
    mkdir -p ${working_dir}/filtered/
    #----------------------------------------------------------------------
    # file1=rsid in f_qc2keep is in column 4
   if [ ! -f ${f_qc2keep} ]
    then
     echo 'Error: there is no qc-ed file for '${chr}
    else
     if [ ! -f ${working_dir}/filtered/${f_out} ] && [ ! -f ${working_dir2}/filtered/${f_out} ] 
      then
      echo Running ${chr}
      if [ ! -f ${chr}.snps2keep ]
      then
       sed 's/\t/ /g' ${f_qc2keep} | sed 's/Position/pos/g' | sed 's/A1/a_0/g' | sed 's/A2/a_1/g' > ${chr}.snps2keep
      fi
      wc ${chr}.snps2keep # 740942 
      # file2=rsid in f (association output) is in column 2
      ##zless ${f}
      # chr rsid pos a_0 a_1 af info 
      zcat ${working_dir}/${f} > ${f2}.tmp
      # subset association output so that it only contains QCd snps within snps2keep, and add snp QC columns
      # it's not keeping columns from file 1,... but does the subsetting at least, will combine with f_qc2keep in R to add these columns in a more controlled way
      # a[$x] in the second {} needs to refer to the relevant column in file 2. I initially called a[$4] and that failed.
      ## check that the selected columns match:
      head ${chr}.snps2keep | awk '{print $4,$3,$5,$6}'
      head ${f2}.tmp | awk '{print $2,$3,$4,$5}'
      # combine
      awk 'NR==FNR{a[$4,$3,$5,$6]=$0;next}($2,$3,$4,$5) in a{print $0,a[$2,$3,$4,$5]}' ${chr}.snps2keep ${f2}.tmp | gzip > ${working_dir}/filtered/${f_out}
      rm ${f2}.tmp ${chr}.snps2keep
     fi
   fi
  fi
done

